src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/rc_app_extension.cc

```diff
@@ -81,14 +122,109 @@ bool RCAppExtension::IsSubscribedToInteriorVehicleData(
 }
 
 void RCAppExtension::SaveResumptionData(
-    smart_objects::SmartObject& resumption_data) {}
+    smart_objects::SmartObject& resumption_data) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (subscribed_interior_vehicle_data_.empty()) {
+    LOG4CXX_DEBUG(logger_, "Subscribed modules data is absent");
+    return;
+  }
+
+  resumption_data[message_params::kModuleData] =
+      smart_objects::SmartObject(smart_objects::SmartType_Array);
+  auto& module_data = resumption_data[message_params::kModuleData];
+
+  uint32_t index = 0;
+  for (const auto& module_uid : subscribed_interior_vehicle_data_) {
+    LOG4CXX_DEBUG(logger_,
+                  "Save module: [" << module_uid.first << ":"
+                                   << module_uid.second << "]");
+    auto module_info_so =
+        smart_objects::SmartObject(smart_objects::SmartType_Map);
+    module_info_so[message_params::kModuleType] = module_uid.first;
+    module_info_so[message_params::kModuleId] = module_uid.second;
+    module_data[index++] = module_info_so;
+  }
+}
 
 void RCAppExtension::ProcessResumption(
     const smart_objects::SmartObject& saved_app,
-    resumption::Subscriber subscriber) {}
+    resumption::Subscriber subscriber) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_TRACE(logger_, "app id : " << application_.app_id());
+
+  if (!saved_app.keyExists(
+          application_manager::strings::application_subscriptions)) {
+    LOG4CXX_DEBUG(logger_, "application_subscriptions section does not exist");
+    return;
+  }
+
+  const auto& resumption_data =
+      saved_app[application_manager::strings::application_subscriptions];
+
+  if (!resumption_data.keyExists(message_params::kModuleData)) {
+    LOG4CXX_DEBUG(logger_, "kModuleData section does not exist");
+    return;
+  }
+
+  auto& module_data = resumption_data[message_params::kModuleData];
+
+  if (0 == module_data.length()) {
+    LOG4CXX_WARN(logger_, "Subscribed modules data is absent");
+    return;
+  }
+
+  for (const auto& module_so : *module_data.asArray()) {
+    const auto module_type = module_so[message_params::kModuleType].asString();
+    const auto module_id = module_so[message_params::kModuleId].asString();
+
+    ModuleUid module{module_type, module_id};
+    LOG4CXX_TRACE(logger_,
+                  "app id " << application_.app_id() << " type : "
+                            << module_type << " id <" << module_id);
+    SubscribeToInteriorVehicleData(module);
+  }
+
+  plugin_.ProcessResumptionSubscription(application_, *this, subscriber);
+}
 
 void RCAppExtension::RevertResumption(
-    const smart_objects::SmartObject& subscriptions) {}
+    const smart_objects::SmartObject& subscriptions_so) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  subscribed_interior_vehicle_data_.clear();
```

Comment:
@jacobkeeler 
I believe we don't need UpdateHash here because we call UpdateHash during ResumptionDataProcessor::ProcessResponseFromHMI in case resumption failed.

Reply:
@IGapchuk I'm not seeing an `OnHashChange` call in `ProcessResponseFromHMI`

Reply:
@jacobkeeler Yes, sure.
We can go through the next sequence of function calls:

1. During application registration, SDL starts the resumption process. Creates callback with internal application hash updating and passes it into ResumeController:
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc#L778

2. The resume controller starts to restore application data and passes taken callback farther:
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resume_ctrl_impl.cc#L434

3. The resume controller passes callback to the resumption data processor for restoring application data:
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resume_ctrl_impl.cc#L649

4. Resumption data processor saves callback in pair with app_id:
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resumption_data_processor.cc#L106

5. During processing response from HMI, the resumption data processor gets callback by app_id and executes it for both cases (Resumption success/Resumption unsuccess):
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resumption_data_processor.cc#L376
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resumption_data_processor.cc#L385
https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resumption_data_processor.cc#L389

Reply:
With regards to this, it seems like the order of the calls needs to be changed, currently the order [here](https://github.com/LuxoftSDL/sdl_core/blob/feature/sdl_0188_interior_vehicle_data_resumption/src/components/application_manager/src/resumption/resumption_data_processor.cc#L381) is 
```
    SDL_LOG_ERROR("Resumption for app " << app_id << " failed");
    callback(mobile_apis::Result::RESUME_FAILED, "Data resumption failed");
    RevertRestoredData(application_manager_.application(app_id));
    application_manager_.state_controller().DropPostponedWindows(app_id);
```
It seems to me that the `callback` should be sent after the `RevertRestoredData` call (and maybe the `DropPostponedWindows` call as well). This would ensure that the hash is updated after all of the data is reverted.

Reply:
@jacobkeeler We believe that is redundant to change sequence of function calls.
The resumption reverting could spend enough long time and mobile device will await along all resumption reverting processes. Also, the transport SDL-HMI and transport SDL-Mobile Device are different transports and 
we can't be sure about exactly the RPCs sequence between SDL, Mobile, and HMI.
We believe for improve UX SDL should first of all notify mobile device about failed resumption and after that start to revert it.